Jake Jsonparser step2

ChocOZerO opened the pull request:
객체가 입력될 경우 처리

ChocOZerO committed:
[Add]객체를 Dictionary타입으로 저장

객체가 Dictionary 타입으로 저장되도록 추가

ChocOZerO committed:
[Add]객체와 아닌 것 구분해서 JSONData에 넣기

String 의 Extension을 따로 파일을 만들어서 관리
객체인지 아닌지 "{"로 구분해서 처리

ChocOZerO committed:
[Update]객체인 경우 출력

ChocOZerO committed:
[Update]JSONObject 타입 추가해서 적용

Dictionary<String, Any> -> JSONObject

godrm commented on JSONParser/JSONParserTests/JSONAnalyserTest.swift:
```diff
@@ -40,7 +40,7 @@ class JSONAnalyserTest: XCTestCase {
         let dic : Dictionary<String, Any> = objectJSONData[0] as! Dictionary<String, Any>
         XCTAssertEqual(dic.count, 4)
```
여기처럼 테스트 케이스가 두 개면 테스트 함수도 분리하세요.

ChocOZerO committed:
[Update]테스트케이스 분리

ChocOZerO commented:
테스트 케이스 분리했습니다.

godrm commented on JSONParser/JSONParser/OutputView.swift:
```diff
@@ -12,13 +12,17 @@ struct OutputView {
         }
         result += getIntCounter()
-        if (typeCounter.boolCounter > 0) && ((typeCounter.stringCounter > 0) || (typeCounter.intCounter > 0)) {
+        if (typeCounter.boolCounter > 0) && ((typeCounter.stringCounter > 0) || (typeCounter.intCounter > 0) || (typeCounter.objectCounter > 0)) {
```
이런 if 문은 중복인 것 같은데 이 방법 뿐일까요? 개선해보세요.

ChocOZerO committed:
[Update]출력시 불필요한 중복코드 제거

ChocOZerO commented:
불필요한 부분 코드 줄였습니다.

godrm commented on JSONParser/JSONParser/OutputView.swift:
```diff
@@ -12,38 +12,42 @@ struct OutputView {
     private(set) var typeCounter : TypeCounter
```
OutputView 내부 typeCounter 속성이 외부에서 접근 가능해야 하는 건가요?

ChocOZerO committed:
[Update]TypeCounter total 구하는 방식 변경

기존에 프로퍼티로 관리하던 total을 메소드 방식으로 변경
출력시에 OutputView에 초기화 시키는 TypeCounter 접근제한 변경
private(set) -> priavte

ChocOZerO commented:
접근 제한자 수정하면서 TypeCounter의 total 구하는 방식 변경했습니다.

godrm commented on JSONParser/JSONParser/TypeCounter.swift:
```diff
@@ -12,22 +12,49 @@ struct TypeCounter {
     private(set) var totalCounter : Int = 0
+    private(set) var container : String = "배열"
     
     mutating func countTypes(items: JSONData) {
```
TypeCounter라는 객체는 생성자를 따로 만들지 않아서 초기값만으로 만들어질 수 있는 객체인데
엄밀히 보면 countTypes(items:) 메소드를 호출해야만 카운터가 들어가고 의미있는 데이터가 만들어집니다.
그래서 제 생각에는 countTypes() 메소드가 명시적인 생성자 역할을 하는 것 같습니다. 어떻게 생각하세요?

ChocOZerO replied:
명시적인 생성자라는건 팩토리패턴 방식으로 만들 수 있을거 같은데요, 이 경우는 생성자로 변경하는게 좋을까요? 아니면 팩토리패턴처럼 함수명을 그대로 차용하고 init을 숨기는게 좋을까요?
그리고 해당 메서드 안에 강제 타입변형이 있어서 codacy에서 이슈로 잡던데 오전에 알려주신 if let 방식으로 변경하는 방법이랑 비교하면 어떤게 더 나을까요?

godrm replied:
제 생각에는 init(items:) 형태여도 충분할 것 같았습니다.

godrm replied:
어차피 is로 확인하고 as!를 하는거라 큰 문제는 없어 보입니다만,
모든 객체마다 타입을 확인하는 구조는 나쁜 냄새가 있네요.

ChocOZerO committed:
[Refactoring]의미와 용도에 맞게 변수명, 함수명 등 수정

+ TypeCounter에 countTypes를 이니셜라이저로 수정
인스턴스 생성과 동시에 결과 도출

+ JSONAnalyser의 메소드들을 타입메서드로 변경
인스턴스 생성 없이 바로 분석하여 JSONData 도출

+ 테스트코드 정리

ChocOZerO commented:
좀 더 의미와 용도에 맞는 코드구조라 생각하며 리팩토링 했습니다.

godrm commented on JSONParser/JSONParserTests/JSONAnalyserTest.swift:
```diff
@@ -21,11 +25,11 @@ class JSONAnalyserTest: XCTestCase {
     func testGetJSONData() {
-        let tester : String = "[ 10, \"jk\", 4, \"314\", 99, \"crong\", false ]"
-        let jsonData = jsonAnalyser.getJSONData(items: tester.getElementsAll())
+        jsonData = jsonAnalyser.getJSONData(items: tester.getElementsAll())
```
이런 경우는 어디서 오류가 있을지 모르기 때문에 tester.getElementsAll() 함수도 별도 테스트로 분리해주는 게 좋습니다.

ChocOZerO replied:
해당 함수에 대한 테스트는 JSONStringExtension에서 따로 구현했는데 이 부분에서도 추가해야되는건가요??

godrm replied:
그렇군요. 다른 함수도 변경 가능성이 있으니까 의존적이지 않는게 좋습니다.
getElementsAll() 리턴 타입을 예측해서 여러 경우를 테스트하는게 더 좋죠.

ChocOZerO committed:
[Update]JSON데이터 테스트 케이스 개선

ChocOZerO commented:
테스트 케이스 개선했습니다.

godrm commented:
수고하셨습니다. codacy도 해결해보세요 :)
